src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/JsonReaderExtensions.cs

```diff
@@ -53,5 +54,43 @@ public static string ReadNextStringProperty(this JsonReader reader, string prope
 
             throw new JsonSerializationException("Could not find string property '" + propertyName + "'.");
         }
+
+        public static IReadOnlyList<T> ReadPropertyArray<T>(this JsonReader reader, JsonSerializer serializer, string expectedPropertyName)
+            where T : class
+        {
+            // Ensure we can read the property name
+            if (!reader.ReadTokenAndAdvance(JsonToken.PropertyName, out var propertyName) ||
+                propertyName.ToString() != expectedPropertyName)
+            {
+                return Array.Empty<T>();
+            }
+
+            // Ensure we're at the start of an array
+            if (!reader.ReadTokenAndAdvance(JsonToken.StartArray, out _))
+            {
+                return Array.Empty<T>();
+            }
+
+            var results = new List<T>();
+
+            do
+            {
+                var result = serializer.Deserialize<T>(reader);
+
+                if (result != null)
+                {
+                    results.Add(result);
+                }
+
+                if (reader.TokenType == JsonToken.EndObject)
```

Comment:
Interesting, does the serializer.Deserialize not handle the EndObject token?

Reply:
In a generic sense, I would expect it to. 

In practice, we don't in certain situations ([example](https://github.com/dotnet/aspnetcore-tooling/blob/06848c4725e55aff10972241801f69a184cca4c1/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs#L75-L77)). I played around with adding the end token reading there, but that led to some other issues, which made me suspect the omission may be by design? Just added this optional end token reading here as a safeguard.